### PR TITLE
vo_gpu_next: add --libplacebo-preset

### PR DESCRIPTION
**Pull request by kasper93**

### CODE REVIEW

**Comment by haasn** [comment]
technically we also have the `fast` preset, it maybe makes more sense to parse this as an enum `fast|default|highquality`

**Comment by haasn** [comment]
I'm not sure I'm happy with this, it leaves a bit of a sour taste in my mouth because it seems to compete with mpv's own profile system and settings. If we're in the business of mapping every option under the sun anyway, then this is redundant. It would make more sense if we _didn't_ map every libplacebo option and only provided a high-level quality select. Which would definitely carry some advantages with it, e.g. the ability to change defaults in only one project (rather than several). But that is not the world we live in, currently.

**Comment by hooke007** [comment]
I thought it was a profile, but after checking the commits I am quite confused.
What it does and why it is not profile?

**Comment by kasper93** [comment]
> I'm not sure I'm happy with this, it leaves a bit of a sour taste in my mouth because it seems to compete with mpv's own profile system and settings.

If anything, improvement over current status quo would be immediate availability of new defaults inherited from libplacebo. Which allows to test new features, without the need to duplicate each default value on mpv side. Which in turn could be done in bulk, once updating minimal libplacebo version supported, instead doing it ever other day for every single float libplacebo adds.

Also I disagree that it competes with mpv's own profiles/parameters. mpv's settings always takes precedence. If anything is not exposed, will be initialized correctly, else will be set to mpv's value.

> If we're in the business of mapping every option under the sun anyway, then this is redundant.

That's true, but we are often lagging behind. Although it is easier to forget to update mpv, than actually do it. So it's probably not a problem.

> It would make more sense if we _didn't_ map every libplacebo option and only provided a high-level quality select. Which would definitely carry some advantages with it, e.g. the ability to change defaults in only one project (rather than several). But that is not the world we live in, currently.

Sure we are not there. How, about we move into this world, with baby steps?

mpv is highly customizable, with every single settings available to the user. But the honest truth is users does not understand what does those options mean. Most of them is copying some config from some random guy on the internet and hope it does what they want. And then you have not only libplacebo defaults, mpv defaults, but "random config" defaults, which is not updated for months, but everyone still uses it, because it is "recommended".

I agree that this presets in its current state does not change much if anything, because options are overwritten anyway, but in the same time I will probably keep it locally just to inherit new libplacebo options, as they come.

**Comment by kasper93** [comment]
> I thought it was a profile, but after checking the commits I am quite confused.
What it does and why it is not profile?

Because we don't need more profiles, I think.

And what it does it inherits parameters form libplacebo library, specifically those that are not mapped in mpv options.

But I hear you, we will probably not merge this ever two unhappy voices is enough for me to give up on the presets idea.

**Comment by haasn** [comment]
I think it has merit, just needs a different coat of paint. I was thinking `libplacebo-preset=<fast|default|highquality>`, with `default` being the default, and `highquality` being set in `profile=gpu-hq`.

This way it should hopefully be clearer to everybody, what's going on.

**Comment by hooke007** [comment]
Still confused.

> if we didn't map every libplacebo option

Actually, I doubt mapping vo_gpu's options might not be a good idea for future.
Some options though share the same name but show different output. It makes users feel confused and think it is a bug.

It means gpu-next should use its own options and build a standalone page `libplacebo's option` in manual (like https://mpv.io/manual/master/#gpu-renderer-options). Of course it would also generate much dup contents because we all know most options do the similar work.

**Comment by kasper93** [comment]
> It means gpu-next should use its own options and build a standalone page libplacebo's option in manual (like https://mpv.io/manual/master/#gpu-renderer-options). Of course it would also generate much dup contents because we all know most options do the similar work.

That would be too much separation. libplacebo as gpu-next is tightly integrated already with doing it's best to mimic vo_gpu. Of course as libplacebo is developed more and more features/options diverge. The real work to make this differences go away would be to slowly deprecate vo_gpu.

**Comment by kasper93** [comment]
> I think it has merit, just needs a different coat of paint. I was thinking libplacebo-preset=<fast|default|highquality>, with default being the default, and highquality being set in profile=gpu-hq.

Done. I will let you decide, how useful it is.

**Comment by haasn** [comment]
LGTM now, I will let others have a comment first

File: video/out/vo_gpu_next.c [hunk 1]
```diff
@@ -56,6 +56,12 @@
 #include "osdep/windows_utils.h"
 #endif
 
+enum quality_preset {
+    PRESET_FAST,
+    PRESET_DEFAULT,
```
Comment (sfan5) [inline video/out/vo_gpu_next.c:61]: very minor but for consistency might be useful to have `PRESET_DEFAULT == 0`

Comment (kasper93) [reply video/out/vo_gpu_next.c:60]: sure, I sorted it by the speed/features enabled, but 0 as default makes more sense.

**Comment by haasn** [comment]
> I guess mpv behavior will first use the default values, then if user used profile=gpu-hq it will use scale=spline36 and then it will detect that libplacebo-preset=highquality and will change that to scale=ewa_lanczos

No, `libplacebo-preset` explicitly only affects options *not* exposed by mpv. So for any option which already exists as an mpv option, the mpv option will be used, end of story.

As of the current status quo I don't think this PR needs to be merged until another pressing use case emerges, currently (with the introduction of `--hdr-peak-percentile`) there is no difference.

**Comment by kasper93** [comment]
As we all have mixed feeling about this change, I will close for now and can be resurrected if needed.

I think the change itself doesn't harm, but currently does not bring any value either, so it is hard to justify it, by possible future needs.

**Pull request closed without merging**